aten/src/ATen/core/DistributionsHelper.h

```diff
@@ -0,0 +1,328 @@
+#pragma once
+
+#include <ATen/CPUGenerator.h>
+#include <type_traits>
+
+// Distributions kernel adapted from THRandom.cpp
+// The kernels try to follow std::random distributions signature
+// For instance: in ATen
+//      CPUGenerator* gen = new CPUGenerator();
+//      at::uniform_real_distribution<double> uniform(0, 1);
+//      auto sample = uniform(gen);
+//      
+//      vs std::random
+//
+//      std::mt19937 gen;
+//      std::uniform_real_distribution uniform(0, 1);
+//      auto sample = uniform(gen);
+//
+// Note: Why are operator() signatures different for CUDA and CPU?
+//       This is because, for CUDA, we will be using Philox4_32_10 engine
+//       inside a kernel and the number of random samples we would need is known
+//       before hand. For the CPU side, it's easier to let CPUGenerator encapsulate
+//       Philox4_32_10 and manage its state. Hence, for CUDA, operator() takes
+//       the Philox4_32_10 object directly and CPU take CPUGenerator object directly.
+namespace at {
+
+/** 
+ * Samples a uniform distribution in the range [0,1] of type T
+ */
+template <typename T>
+struct uniform_real_distribution {
+
+  C10_HOST_DEVICE inline uniform_real_distribution(T a_in, T b_in) {
+    #ifndef __CUDA_ARCH__
+      AT_ASSERT(a_in <= b_in);
+      AT_ASSERT(b_in-a_in <= std::numeric_limits<T>::max());
+    #endif
+    a = a_in;
+    b = b_in;
+  }
+
+  template<typename U = T> 
+  C10_DEVICE inline typename std::enable_if<std::is_same<U, double>::value, double>::type
+  operator()(Philox4_32_10& engine){
+    uint64_t hi = static_cast<uint64_t>(engine()) << 32;
+    uint64_t lo = static_cast<uint64_t>(engine());
+    uint64_t random = hi | lo;
+    double x = static_cast<double>(random) / static_cast<double>(std::numeric_limits<uint64_t>::max());
+    return (x * (b - a) + a);
+
+  }
+
+  template<typename U = T> 
+  C10_DEVICE inline typename std::enable_if<std::is_same<U, float>::value
+                                          || std::is_same<U, at::Half>::value, float>::type
+  operator()(Philox4_32_10& engine){
+    float x = static_cast<float>(engine()) / static_cast<float>(std::numeric_limits<uint32_t>::max());
```

Comment:
so for a=0, b=1 this is going to be 0-inclusive 1-inclusive?  It is also going to break in box-muller then when you are trying to do log(1-u)

Reply:
How about this method then: `float x = static_cast<int>(engine()) * INV_2_POW_32 + 0.5;` which gives [0,1)
from: [http://www.doornik.com/research/randomdouble.pdf]()
```
The unsigned random integer is cast to a signed integer (which is costless), and
multiplied by 2^32. This gives a value in [−0.5, 0.5) to which 0.5 must be added.
The values are now generated in a different order: for the unsigned integer running
from 0 to the maximum, the uniform values run from 0.5 to 0.999999999767169 and
then from 0.0 to 0.4999999997671694......it is costless to create a version that
excludes zero, by adding 2−32/2 to all values (half the step size of the grid of
numbers).
```
CURAND pretty much takes this approach except for doing `x * CURAND_2POW32_INV + (CURAND_2POW32_INV/2.0f);` which is on (0,1) interval.

My soul is pretty dead thinking about this random int to uniform floats conversion... :laughing:

Reply:
`uniform` is supposed to produce 0-inclusive 1-exclusive distribution https://github.com/pytorch/pytorch/issues/2740.  The method above won't work for 1-exclusive part - INT_MAX/INV_2_POW_32 is 0.5 exactly, not just under 0.5.  
```
In [19]: a[0] = ((2**31-1)/2**32)                                                                                                              

In [20]: a                                                                                                                                     
Out[20]: tensor([0.5000])
```
However, even suppose it worked, and you then directly reuse this uniform in box-muller, you'll break on log.